smithy-model/src/main/java/software/amazon/smithy/model/transform/FlattenPaginationInfo.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.model.transform;
+
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.PaginatedIndex;
+import software.amazon.smithy.model.knowledge.PaginationInfo;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.traits.PaginatedTrait;
+
+/**
+ * Flattens pagination info from service shapes into operation-level pagination traits.
+ */
+final class FlattenPaginationInfo {
+
+    private final ServiceShape service;
+
+    FlattenPaginationInfo(ServiceShape service) {
+        this.service = service;
+    }
+
+    public Model transform(ModelTransformer transformer, Model model) {
+
+        Optional<PaginatedTrait> serviceLevelPagination = service.getTrait(PaginatedTrait.class);
+        if (!serviceLevelPagination.isPresent()) {
+            return model;
+        }
+        PaginatedIndex paginatedIndex = PaginatedIndex.of(model);
+
+        // Merge service-level information into each operation's pagination trait.
+        Set<Shape> updatedShapes = new HashSet<>();
+        for (OperationShape operationShape : model.getOperationShapes()) {
+            Optional<PaginationInfo> infoOptional = paginatedIndex.getPaginationInfo(service, operationShape);
```

Comment:
A non-empty Optional from `getPaginatedInfo` requires the operation to have the `@paginated` trait applied, meaning the loop could look for that trait (`getOperationShapesWithTrait()`) and the pre-checking can be removed, simplifying the rest of this loop.